Automattic/wp-calypso
# Simple payments: redux data layer

This leverages custom post type `jp_pay_product` to provide product management functionality for Simple Payments.
Uses data layer.

Works with https://github.com/Automattic/jetpack/pull/7409

- Introduces data layer to communicate with /posts?type=jp_pay_product
- Introduces simple-payments redux subtree
- Introduces actions for tying this communication together
- Introduces schemas for sanitizing / translating responses to product models

## Intentionally NOT covered in this PR:

This is a monster already. Purpose of this is to unblock calypso work ASAP

- Tests (feel free to add)
- simple action creators (they should be a breeze to make)
- requesting statuses for edit / delete / add. We should be able to work without them for now since items list is always updated. I dont think we need these requesting statuses.

## Usage

### Adding new product

```
dispatch({
	type: 'SIMPLE_PAYMENTS_PRODUCTS_LIST_ADD',
	siteId: 118642378,
	product: {
		title: 'new product',
		description: 'best new product ever',
		price: '110.98',
		currency: 'PLN',
		status: 'enabled',
		multiple: 1,
		}
});
```

### Editing product

```
dispatch({
	type: 'SIMPLE_PAYMENTS_PRODUCTS_LIST_EDIT',
	siteId: 118642378,
	productId: 87,
	product: {
		title: 'really old product',
		description: 'best new product ever',
		price: '12',
		}
});
```

### Deleting a product

```
dispatch({
type: 'SIMPLE_PAYMENTS_PRODUCTS_LIST_DELETE',
siteId: 118642378,
productId: 87,
});
```

### Listing product list

```
dispatch({
type: 'SIMPLE_PAYMENTS_PRODUCTS_LIST',
siteId: 118642378
});
```

# Testing

- Pick a JP site
- Switch to this branch of JP https://github.com/Automattic/jetpack/pull/7409
- checkout this calypso buld
- you can run `dispatch` as a global in console on development
- run add product few times, see if list populated
- list product, see if list is populated
- edit a product, see if it is edited in redux tree
- delete a product, see if deleted in redux tree
- run product list once again, see if it is correct.


Works with https://github.com/Automattic/jetpack/pull/7409
